### PR TITLE
Add KafkaSource Form 

### PR DESCRIPTION
**Pull request by divyanshiGupta**
Jira - https://issues.redhat.com/browse/ODC-2662

**Analysis / Root cause:**
User can't create KafkaSource via web console

**Solution Description:**
Adds support for creation of KafkaSource via web console

**Screen shots / Gifs for design review:**
kafkaSource

KafkaSource CRD - https://github.com/knative/eventing-contrib/blob/master/kafka/source/config/300-kafkasource.yaml
API - https://github.com/knative/eventing-contrib/blob/master/kafka/source/pkg/apis/sources/v1alpha1/kafka_types.go

**For PR testing setup follow:**
https://knative.dev/docs/eventing/samples/kafka/

To-do

- [x] fix tests

- [x] add validations

- [x] add unit tests

**Unit test coverage report:**
image
image
image

@openshift/team-devconsole-ux

### CODE REVIEW

File: frontend/packages/knative-plugin/src/components/add/event-sources/ServiceAccountDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      showBadge
+    />
+  );
+};
+
+const mapStateToProps = (state: RootState): StateProps => ({
+  namespace: getActiveNamespace(state),
+});
+
+export default connect(mapStateToProps)(ServiceAccountDropdown);
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/components/add/event-sources/ServiceAccountDropdown.tsx:50]: nit: can have type here

File: frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts [hunk 2]
```diff
@@ -18,6 +18,7 @@ import * as camelSourceImg from '../imgs/logos/camelsource.svg';
 import * as containerSourceImg from '../imgs/logos/containersource.png';
 import * as cronJobSourceImg from '../imgs/logos/cronjobsource.png';
 import * as kafkaSourceImg from '../imgs/logos/kafkasource.svg';
+import * as _ from 'lodash';
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts:21]: nit: we can have it above relative imports

File: frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts [hunk 3]
```diff
@@ -77,6 +100,27 @@ export const getEventSourceData = (source: string) => {
         },
       },
     },
+    kafkasource: {
+      spec: {
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts:104]: can we have `resources` and `requests` as well under this spec with default initial values instead of having under `limits` as these values are used only in Kafka Source so am not sure if want to have it separate and with this we can avoid https://github.com/openshift/console/pull/4899/files#diff-745c0439e378574e2e12727d3caa5522R41-R61

WDYT?

File: frontend/packages/knative-plugin/src/components/add/EventSourceForm.tsx [hunk 4]
```diff
@@ -32,11 +34,13 @@ const EventSourceForm: React.FC<FormikProps<FormikValues> & OwnProps> = ({
     <EventSourcesSelector eventSourceList={useEventSourceList()} />
     {values.type === EventSources.CronJobSource && <CronJobSection />}
     {values.type === EventSources.SinkBinding && <SinkBindingSection />}
+    {values.type === EventSources.KafkaSource && <KafkaSourceSection />}
     <SinkSection namespace={namespace} />
     <AppSection
       project={values.project}
       noProjectsAvailable={projects?.loaded && _.isEmpty(projects.data)}
     />
+    <AdvancedSection />
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/components/add/EventSourceForm.tsx:43]: am not sure if we need `<AdvancedSection />` always  as none of the other sources has so maybe can have only for `kafkaSource` or shall we show limits (resources/requests) in main form itself 

cc @christianvogt @serenamarie125

File: frontend/packages/knative-plugin/src/components/add/event-sources/SecretKeySelector.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import * as React from 'react';
+import { k8sGet } from '@console/internal/module/k8s';
+import { SecretModel } from '@console/internal/models';
+import { errorModal } from '@console/internal/components/modals/error-modal';
+import { FormGroup } from '@patternfly/react-core';
+import { ValueFromPair } from '@console/internal/components/utils/value-from-pair';
+import { getFieldId } from '@console/shared';
+import { useFormikContext, FormikValues, useField } from 'formik';
+import { connect } from 'react-redux';
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/components/add/event-sources/SecretKeySelector.tsx:9]: nit: we can have it above internal package imports

File: frontend/packages/knative-plugin/src/components/add/event-sources/SecretKeySelector.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      />
+    </FormGroup>
+  );
+};
+
+const mapStateToProps = (state: RootState): StateProps => ({
+  namespace: getActiveNamespace(state),
+});
+
+export default connect(mapStateToProps)(SecretKeySelector);
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/components/add/event-sources/SecretKeySelector.tsx:65]: nit: can have type here

File: frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts (hunk 3 above)
Comment (divyanshiGupta) [reply frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts:104]: If you see the way ResourceLimitSections is written it requires limit value, unit, default unit etc so we need to have all that info in the form itself. It made sense to keep it outside and then modify it according to our needs in a similar way we do in other import forms.

File: frontend/packages/knative-plugin/src/components/add/EventSourceForm.tsx (hunk 4 above)
Comment (divyanshiGupta) [reply frontend/packages/knative-plugin/src/components/add/EventSourceForm.tsx:43]: If none of the other sources are going to have any advanced options then we can show it only for kafka

**Comment by invincibleJai** [comment]
@serenamarie125 PTAL , let us know if it looks good. 

@divyanshiGupta looks like it needs a rebase

**Comment by serenamarie125** [comment]
this is looking good @divyanshiGupta!  I'm assuming that the Application name piece changes as in other forms ?

**Comment by divyanshiGupta** [comment]
> this is looking good @divyanshiGupta! I'm assuming that the Application name piece changes as in other forms ?

@serenamarie125 yes its the same as other forms. If application groups already exists, a dropdown is shown otherwise a text input field.

**Comment by invincibleJai** [comment]
Verified locally, works as expected

/approve

File: frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts [hunk 7]
```diff
@@ -26,15 +27,37 @@ export const getEventSourcesDepResource = (formData: any): K8sResourceKind => {
...
-    type === EventSources.ApiServerSource || type === EventSources.SinkBinding
+    type === EventSources.ApiServerSource ||
+    type === EventSources.SinkBinding ||
+    type === EventSources.KafkaSource
       ? KNATIVE_EVENT_SOURCE_APIGROUP
       : KNATIVE_EVENT_SOURCE_APIGROUP_DEP;
   const apiVersion = 'v1alpha1';
   const eventSrcData = data[type.toLowerCase()];
+  const completeEventSrcData =
+    type === EventSources.KafkaSource && eventSrcData
```
Comment (invincibleJai) [inline frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts:43]: am not sure if we can avoid this but wanted to have this utils without anyModel check. Working on eliminating above for apiGroup as well.

**Comment by invincibleJai** [comment]
/lgtm

Verified locally, works as expected

File: frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts (hunk 7 above)
Comment (divyanshiGupta) [reply frontend/packages/knative-plugin/src/utils/create-eventsources-utils.ts]: I have refactored the utils and now I have created a base util that returns the common/base resource and event source specific utils to modify the extend the base resource as needed. https://github.com/openshift/console/pull/4899/commits/23149f5c30fad6039f6e491ffb1f7494e23f0c26.

File: frontend/public/module/k8s/k8s.ts [hunk 8]
```diff
@@ -47,7 +47,7 @@ export const referenceForOwnerRef = (ownerRef: OwnerReference): GroupVersionKind
   )(ownerRef.kind);
 
 export const referenceForModel = (model: K8sKind): GroupVersionKind =>
-  referenceForGroupVersionKind(model.apiGroup || 'core')(model.apiVersion)(model.kind);
+  referenceForGroupVersionKind(model?.apiGroup || 'core')(model?.apiVersion)(model?.kind);
```
Comment (christianvogt) [inline frontend/public/module/k8s/k8s.ts:50]: Discussed with @Divyansh42.
Let's revert this change and address the issue in `ResourceDropdown`

Comment (divyanshiGupta) [reply frontend/public/module/k8s/k8s.ts]: Done.
